src/datatable/datatable.jl

```diff
@@ -834,35 +805,12 @@ function Base.convert(::Type{DataTable}, A::Matrix)
     return DataTable(cols, Index(gennames(n)))
 end
 
-function _datatable_from_associative(dnames, d::Associative)
-    p = length(dnames)
-    p == 0 && return DataTable()
-    columns  = Vector{Any}(p)
-    colnames = Vector{Symbol}(p)
-    n = length(d[dnames[1]])
-    for j in 1:p
-        name = dnames[j]
-        col = d[name]
-        if length(col) != n
-            throw(ArgumentError("All columns in Dict must have the same length"))
-        end
-        columns[j] = NullableArray(col)
-        colnames[j] = Symbol(name)
-    end
-    return DataTable(columns, Index(colnames))
-end
-
 function Base.convert(::Type{DataTable}, d::Associative)
-    dnames = collect(keys(d))
-    return _datatable_from_associative(dnames, d)
-end
-
-# A Dict is not sorted or otherwise ordered, and it's nicer to return a
-# DataTable which is ordered in some way
-function Base.convert(::Type{DataTable}, d::Dict)
-    dnames = collect(keys(d))
-    sort!(dnames)
-    return _datatable_from_associative(dnames, d)
+    colnames = collect(keys(d))
+    isa(d, Dict) && sort!(colnames)
```

Comment:
Could keep the comment from above and move it before this line. Also you only need to call `collect` in that branch, so move it inside `sort!`.